app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,424 @@
+  @BindView(R.id.text_view_device_name) TextView deviceName;
+  @BindView(R.id.progress_bar_searching_peers) ProgressBar searchingPeersProgressBar;
+  @BindView(R.id.list_peer_devices) ListView listViewPeerDevices;
+  @BindView(R.id.text_view_empty_peer_list) TextView textViewPeerDevices;
+  @BindView(R.id.recycler_view_transfer_files) RecyclerView filesRecyclerView;
+
+  private boolean isFileSender = false;    // Whether the device is the file sender or not
+
+  private @NonNull WifiDirectManager wifiDirectManager = new WifiDirectManager(this);
+
+  private ArrayList<FileItem> filesForTransfer = new ArrayList<>();
+  private FileListAdapter fileListAdapter;
+
+  private List<WifiP2pDevice> availablePeerDevices = new ArrayList<WifiP2pDevice>();
+  private boolean hasSenderStartedConnection = false;
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    KiwixApplication.getApplicationComponent().activityComponent()
+        .activity(this)
+        .build()
+        .inject(this);
+    setTheme(sharedPreferenceUtil.nightMode() ? R.style.Theme_AppCompat_DayNight_NoActionBar : R.style.AppTheme);
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_local_file_transfer);
+    ButterKnife.bind(this);
+
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    Intent filesIntent = getIntent();
+    ArrayList<Uri> fileUriArrayList;
+    fileUriArrayList = filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
+    isFileSender = (fileUriArrayList != null && fileUriArrayList.size() > 0);
+
+    setSupportActionBar(actionBar);
+    actionBar.setNavigationIcon(R.drawable.ic_close_white_24dp);
+    actionBar.setNavigationOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        finish();
+      }
+    });
+
+    listViewPeerDevices.setAdapter(
+        new WifiPeerListAdapter(this, R.layout.row_peer_device, availablePeerDevices));
+
+    if (isFileSender) {
+      for (int i = 0; i < fileUriArrayList.size(); i++) {
+        filesForTransfer.add(new FileItem(getFileName(fileUriArrayList.get(i)), TO_BE_SENT));
+      }
+
+      displayFileTransferProgress(filesForTransfer);
+    }
+
+    wifiDirectManager.createWifiDirectManager(sharedPreferenceUtil, alertDialogShower, filesForTransfer, fileUriArrayList);
+  }
+
+  @OnItemClick(R.id.list_peer_devices)
+  void onItemClick(int position) {
+    /* Connection can only be initiated by user of the sender device, & only when transfer has not been started */
+    if (!isFileSender || hasSenderStartedConnection) {
```

Comment:
these probably should be in WDM. LFTA reports actions and WDM controls logic

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/0f97a2e8e52e144f281c1fef4e6b8f524f4f0537#diff-e1d59ae6e15dcb502e125fb6e243d15dL144)